### DISCUSSION TITLE
New hardware platform for OV

### DISCUSSION THREAD

**Discussion by linuxianer99** [Openvario/meta-openvario#329] [Ideas]
As the A20 SoC seems to make some stability problems, we should think about a new platform ...

**Comment by linuxianer99** [comment]
As far a i know, Pi4 has no LVDS interface ... only the HDMI. 
How should we adapt the displays to it ??

Should we think about the RPi4 CM ??

**Comment by realtimepeople** [comment]
My friend winfried found a sunlight-readable display with all the interfaces HDMI LVDS... for around 350 EUR
https://www.unisystem-displays.com/en/products/lcd-tft/standard/dlh0708-t.html
This one with some Raspi may do the job.
Ask at https://www.ipc-markt.de/

**Comment by DanD222** [comment]
See also https://forum.xcsoar.org/viewtopic.php?f=30&p=23540#p23540

We have two options to connect the current displays:

1. DSI to LVDS bridge
2. DPI to LVDS bridge

I've been looking into the DPI to LVDS option so far using a SN75LVDS83BDGG chip, as I haven't found many readily available DSI to LVDS bridges. _Edit - the chip I thought was a DSI to LVDS really wasn't one; we'd need something like SN65DSI84, which has a delivery time estimate of about one year from now._

The issue with the DPI interface:
- 24 bit mode takes up all of the GPIO pins (we could save a few GPIO pins by driving the display in 18 bit mode). This would mean we'd need to connect all external devices through USB, for example through a FT4232 chip
- Depending on whether the display has VESA or JEIDA mapping, the DPI-LVDS bridge needs to be wired differently, so we're looking at potentially two different PCB versions depending on which display is used (AM-1280800P3TZQW-00H-AMPI is VESA and is in stock at Texim at the moment; CH070OLDLTH-002-CHEF is out of stock at Texim, and I haven't been able to figure out if it's VESA or JEIDA from the datasheet)

RPi4 CM4 is probably a good route, as this would theoretically allow us to produce hardware that is backwards-compatible with the existing metalwork. We'd effectively be re-creating the Cubieboard PCB though, including whatever peripherals we'd want - we'd need to implement a USB hub for example. There's also the issue of updating the software, as the RPi CM4 doesn't use an SD card slot (only the CM4 Lite). So one would need to bring a laptop to the airfield in order to update an OV, unless we go for the CM4 Lite, but I haven't looked into what the particularities of using that may be. 

If we don't want the hardware to be backwards-compatible we could use the RPi4 B board and skip the USB hub implementation, etc.. but since we'd need to design an Adapterboard anyway, we might as well go the CM4 route I think.

**Comment by lordfolken** [comment]
This is just IMO:
* I would avoid building another adapter board. 
* There is plenty HDMI to LVDS bridges available on ebay. 
* Recommend a good hdmi display for the future builds
* Go the commercial HDMI to LVDS route for existing solutions
* Stick with standard RPI instead of CM module
* Design a good flexible housing with several mounting options (inside for the pcbs)

I would concentrate the hardware design skills on where they really matter: A sensor board. Preferably one that can simply be plugged into whatever via USB/Bluetooth/Wifi.

With this approach you would:
 * decouple the upgrade problems
 * extend the audience (Smart phone / Kobo users )
 * You could do a highly integrated sensor board with all the timing and latency issues resolved (i.e. its own processor)
 * "Standard" NMEA out/in
 * Could addstuff like:
 * Digital IO pins
 * USB Charging port (one that actually works..) 
 * Direct audio for Vario

**Comment by DanD222** [comment]
A new Sensorboard's already in progress: https://github.com/DanD222/sensorboard_NewHardware (with some further improvements for a new revision in the pipeline). @hpmax has a couple of prototypes built, but AFAIK he's coding everything by himself, so I'm sure he could use some help, especially with the ESP32 side of things.

**Comment by mihu-ov** [comment]
>I would concentrate the hardware design skills on where they really matter. A sensor board. Preferably one that can simply be plugged into whatever via USB/Bluetooth/Wifi.

Today, I would probably skip the sensorboard and buy a https://xcvario.de/. I think it has all the features mentioned and even comes with an attached LCD and rotary encoder :-)
https://github.com/iltis42/XCVario has hardware design files and software (don´t know if it is complete and for the latest hardware revision).

The one crucial point IMO is the LCD and how to drive it. We considered using HDMI->LVDS converters during the original OV design but at that time these converters were large, LCD specific, had high power consumption and had their own issues. 
Now if there was a sunlight readable LCD with bonded capacitive touch with low power consumption and at a reasonable price and an adapter to drive it from an RPi I would be in :-)

**Comment by realtimepeople** [comment]
We have also created a new Sensor for OpenVario:
https://github.com/larus-breeze/hw_sensor
This one has (pretty successfully) been tested by SteFly in the German Championship ;-)
The firmware will be published under larus-breeze shortly.
![larus_hw_1_0](https://user-images.githubusercontent.com/62793036/188909140-453c50ec-d348-4548-b9de-f9a338025ba6.jpg)

**Comment by linuxianer99** [comment]
Hi All,

as i see all the comments, i think Openvario Next Gen can just be a Linux System with a Display running XCSoar and lots of external sensor connectors ??

We can do a small "display and calculation" device with CAN, USB, Ethernet, etc ..
We have than an similar approach as Garmin did with the G3X ...

Any comments to this direction ??

**Comment by hpmax** [comment]
Few comments:

1) I've come to the conclusion that I think the sensorboard should be gotten rid of in favor of an external vario.  If someone doesn't want an external vario, we can offer a headless one, or perhaps just a single sensorboard that can be mounted into the case optionally.

2) I think Bluetooth 5 and Wifi are important functions to have.  Why Bluetooth 5?  It can support up to 7 simultaneous Bluetooth connections.  I have been having a devil of a time dealing with radio noise, and I'm sure most of the problem is coming from USB.  Less USB is better.  Better shielded USB (at least on OV side) is critical.  How many RJ45/serial ports are really needed?

3) I kind of like the idea of an external multi-input audio amplifier.  It has some downsides.  Namely single point of failure.  As it stands now, we'd have the OV and radio both wanting a speaker.  If we moved the vario external, that would mean either 3 devices or using the audio capability of OV to generate a signal based on the vario -- this is a problem but adds a big failure mode to the audio vario.  Either way, my experience with the audio amp in Stefan's OV hasn't been good, I've blown out two amplifiers.  I gave Stefan an updated design with a different part which I think handles power better, and I switched to an 8 ohm speaker which I think has resolved the problem.  But there are several advantages to an external multi-input audio amp:

   a) If it fails, it's easily replaced.
   b) Design can be optimized for efficiency...  It could also potentially be placed near the speaker itself meaning that you could use a ery short wire run which decreases EMI issues if a Class D amplifier is used.
   c) Amplifier gains could be changed dynamically...  For example, when the radio receiver is on you turn down the volume on the audio vario... or same goes for an airspace warning or a FLARM warning.  

4) I like the Raspberry Pi for its ubiquity...  It's hard to argue with that, however there are downsides.  The RPi4 is power hungry.  To some degree, we can compensate for that with bigger batteries.  LiFePO4 batteries make 20AH very reasonable in most gliders -- but more power consumption is also more heat... And I feel like a glider cockpit is already a pretty hot environment.  I really don't like the extra heat.  It can cause premature failure of every component, including the display.  I'm also not sure what the display options are.  

It would be nice to see some newer/better options than the long defunct PixelQi and the old Texim display.  I don't know how compatible the RPi4 is with some of the newer displays.  A touch screen would also be nice.  DLF0708T (same display as mentioned above, but more stripped down and with an LVDS interface) seems like a good option, bit pricey, but if the performance is what they claim, it might be worth it.  Much bright than the Texim and higher resolution.  No touch interface though.  I've also heard about 7 color e-ink displays.

It would be nice if careful consideration were given to reducing EMI, particularly through USB...  But I doubt any of them do that.  

As a side note: If anyone is good at ESP32S programming and wants to help out, it would be really nice to get some assistance on the programming side for the second gen sensorboard (note: my intent is to update the hardware a bit, and turn this into an external vario).

**Comment by markus2marks** [comment]
Hi guys, 

if you are interested in upgrading your openvario with Raspberry, you can contact me. I have already developed a board that will be replaced by the cubieboard. All the functions of the openvario are retained and you then have the option of connecting an LTE modem via the m.2 data slot on the back. This makes it very easy to get live data into your openvario :) 

![IMG_9672_3](https://github.com/user-attachments/assets/1663533a-cc3b-4424-9713-e864331c0cd4)

**Reply by OBrown92** [reply]
Pretty interesting. For which OV's does it fit? We have the Stefly OV's in our Club but have some issues with crashes. Can we keep the display?

**Comment by markus2marks** [comment]
Yes, that's the idea, just change the cubieboard and leave the rest the same. You can use it for both variants.

**Reply by lordfolken** [reply]
Would love to see it in action. Where can i get the information?

**Reply by markus2marks** [reply]
There is no project site for it at the moment. I simply don't have the time. What do you want to know?

**Reply by lordfolken** [reply]
hmm. That's too bad. :/
I was hoping for something reproducible that the 1500+ openvarios could potentially use as an upgrade path.

**Reply by hpmax** [reply]
Based on talking to Markus, I think getting a "group buy" situation going would really help drive the price down.    I didn't realize there were that many Stefly OVs sold.  I think the main issue is the form factor needs to stay the same.   The other question is there a better display that meets the same form factor.

**Reply by linuxianer99** [reply]
Hey All! 
Thanks to Markus for this great work!
I am also very interested in this development and buying some of these boards.

Maybe we can also use this upgrade to bring some of the previously discussed changes (New version naming, better update) to the OV image.
I will also try to spend again more resources for the OV ...

@markus2marks : Can you share a schematic or PCB layout of your baseboard ?
Think the most critical part is the display connection. Or maybe we can also have a call on this ?

br,
Timo

**Reply by markus2marks** [reply]
yes i can share the schematic and i will setup project site for it in the next days. The display connection is no problem, i have already solved it :)

**Reply by linuxianer99** [reply]
Cool!
My doubts according display are about the 4,3" RGB ... This is not LVDS/HDMI but parallel display lines ..

**Reply by DanD222** [reply]
@markus2marks You can always upload the files to a git repo, like with Larus, for example.

One thing to keep in mind though - here’s a friend’s RPi5 build in a 3D printed case:

https://github.com/user-attachments/assets/6937f85f-594c-4bd3-bd24-1024ef94ed70

**Reply by HoeckDK** [reply]
Looks really interesting. Would be nice future proof the OV. Would be interested to buy a board.

**Reply by markus2marks** [reply]
Good to hear, we are still working on it, maybe you can email me and we can keep in touch :)

**Reply by lordfolken** [reply]
Why dont you upload what you have? Doesnt have to be finished.

**Reply by DanD222** [reply]
(plus we’ll also be able to tell you if we think you shouldn’t be skipping electrolytics around the buck converter before you put the boards into production, and things like that) ;)

**Reply by markus2marks** [reply]
I did not change the buck converter board, only the CB board. I don't know which board you are using in the video. I have the pcb from Stefly.

**Reply by DanD222** [reply]
> I did not change the buck converter board, only the CB board. I don't know which board you are using in the video. I have the pcb from Stefly.

OK - I misunderstood what it was you’re doing.

**Reply by linuxianer99** [reply]
I also did an update to the current version of Yocto ... you can find in the master-next branch ...
But it is not complete yet ...

**Reply by HoeckDK** [reply]
How about the firmware side of things?
Before I went the SteFly OV way, I intended to use a RasPi with HDMI display. That did not work out because the FW had issues. But this was back in 2016-2017.

Is the HDMI bit needed? I know the Cubie has it, but the use of it would be very rare?

Do you have working boards available?

**Reply by markus2marks** [reply]
yes, I'm still building up some boards, but there are still some parts missing, maybe I'll be done with it next week, then you can have one if you want to test it. There is an hdmi port on the board that can be used. But the LCD is connected via the LVDS interface.

**Reply by HoeckDK** [reply]
That would be very interesting. 

Have you already got a compatible SW build going?

**Reply by markus2marks** [reply]
Yesss!!! The latest hardware version is running stable, timo is still setting up the yocto configuration.

**Reply by lordfolken** [reply]
I would be interested in a board.

**Reply by HoeckDK** [reply]
How is the progress on the prototype? I saw that you have uploaded a schematic. It does not look too involved. :)

**Reply by markus2marks** [reply]
Yocto is running, LCD is working, but the cross-compiling of xcsoar is not working at the moment.

**Comment by markus2marks** [comment]
The CM5 module is also working
![image](https://github.com/user-attachments/assets/7bd584fc-0522-4b80-aaa1-64335f8db835)

**Reply by lordfolken** [reply]
Shiny! +1

**Reply by linuxianer99** [reply]
Very cool!!

**Comment by hpmax** [comment]
It's been a few months, is there an update on the hardware/software for this new configuration?  Also, for those of us using SteFly hardware, will we be able to keep all the connectivity the same?  I really believe that ideally the box should effectively just be a tablet with some additional connectivity, and we should rely on an external instrument for the sensorbox.  

@DanD222 and I designed something, which I think is actually pretty good but I've kind of abandoned it because I don't really know what software algorithms to use.  I have an XCVario which I will say is very nicely packaged, but the hardware does leave a lot to be desired.

**Comment by HoeckDK** [comment]
Hi all

Supported by @markus2marks I managed to get the last bits working, that I need: 3rd serial and removing LCD whine. 

So yesterday I installed the CM4 board in my glider and flew it. 

It is a nice upgrade over the cubie. Newer XCsoar and snappier operation. 

I only use it for navigation. So I cannot speak for the vario part. 

Nice project to give the original SteFlys a new heart. Thanks @markus2marks 

Only problem is that the driver for CAI vario seems not compatible with AirGlide task declaration anymore. But that must be a XCsoar thing. Anybody seen the same?

**Comment by OBrown92** [comment]
We flew the CM4 with @markus2marks pcb for one season in our club glider. After some issues with boot loops (probably because of the connected speaker, is disconnected for now), it worked flawless. CB2 crashed super often during flight, now issues with the new board. This should be the way to go for old devices, swap was easy as well. I cut a corner from our stefly housing to reach the usb for programming.

**Comment by lordfolken** [comment]
Yes i second the cm4 module. Can you somehow announce it on the homepage and make an offical ordering process? Maybe stefly could put it into the shop?

**Reply by linuxianer99** [reply]
I am not sure, if Stefan wanna put it in his shop. ... I guess Stefly will support the SteFly Nav (Android) platform more than updating the "old" hardware ...
To be honest, i would like to keep it open source ... Maybe we can create a forum on Openvario.de for ordering some batches ?

@markus2marks Guess. you will use JLCPCB as manufacturer ??

@lordfolken Regarding publish on the homepage ... Would it be fine for you, to have a look at the homepage if i give you admin rights ??

**Reply by markus2marks** [reply]
yes, i use JLCPCB :)

**Comment by linuxianer99** [comment]
Good idea. I am in the process of merging the Opensoar changes with the ney yocto base layers. Linux Variant is working, but pi variant does not build anymore ... Have to look this up ...

**Reply by lordfolken** [reply]
If i broke something, please do tell me any errors you have.

**Comment by HoeckDK** [comment]
I have flown maybe 30 hrs now with the CM4 replacement for the cubie board. Great improvement over the cubie.

There has been some amount of debugging and fixing to get here. But at this point I have 3 bidirectional serials (1 was only RX (CAI driver not working) in the beginning and the other one occupied by uboot, cmdline.txt+config.txt+uboot.env fix), no whine in 7" LCD (ov-rpi4-7-lvds.dtbo bug/fix), and warning sounds coming through (XCsoar https://github.com/XCSoar/XCSoar/pull/2115). That should cover my needs. 

Big thanks to @markus2marks to get here and develop the PCB/image. XCsoar dev team has also been helpful. 

Now I just need the flying season to begin up here:)

**Comment by bomilkar** [comment]
@markus2marks : reading through the above, can you tell me **where I can buy the PCB to replace the Cubie2 by a CM4** (or Pi4)?
I'm running a SteFly without sensorboard, and a 7" CH070 display. I would like to keep the CH070 mainly because I don't want to change the panel.

BTW: a friend of mine is using a Waveshare "7,0 inch IPS Display for Raspberry Pi with DPI Interface" since several years. It works well, sunlight readability is good, it is not power hungry, and it's dirt cheap. But, I'd have to apply changes to my panel if I were to us it.

**WRT the S/W**: since it has no sensor board, all one needs is xcsoar to run on it. Compiling this on the CM4 is just a matter of less than 1 hour. No issues at all, essentially just "git clone" and "make". (Make sure you run "sudo ide/provisioning/install-debian-packages.sh" to install all the right packages before you "make".)

**WRT the H/W**: I'm running xcsoar (with the DPI display) on a "Pi Zero 2 W Rev 1.0". It is good enough for the job. (Compiling on in target takes LONG, but that's not an issue since this board is binary compatible to a Pi4 or CM4.) The advantage of Pi Zero 2 W: there is no heat issue whatsoever! But be careful: a Pi Zero 2 W is VERY different from a Pi Zero, which is a 32 bit CPU and NOT code compatible to a Pi4.

**Reply by markus2marks** [reply]
Send me an email, and we can clarify the details.

**Comment by markus2marks** [comment]
I'm still working on an update version for KiCad:
<img width="833" height="436" alt="grafik" src="https://github.com/user-attachments/assets/9be4c226-fe25-4b5a-b09f-937e543475ed" />
